File: FinMonteCarlo.cpp
#include <iostream>
#include <random>
#include <cmath>
#include <chrono>

using namespace std;

std::default_random_engine generator;

double stoc_walk(double p,double dr,double vol,int periods)
{
    double mean = 0.0;
    double stdv = 1.0;

    std::normal_distribution<double> distribution(mean, stdv);

    for(int i=0; i < periods; i++)
    {
        double w = distribution(generator);
        p += dr*p + w*vol*p;
    }
    return p;
}

int main()
{
    //Initialize variables
    double s0 = 10.0;           //Initial price
    double drift = 0.001502;    //daily drift
    double volatility = 0.026;  //volatility (daily)
    double r = 0.02;            //Risk free yearly rate
    int days = 255;             //Days
    int N = 100000;             //Number of Monte Carlo trials
    double zero_trials = 0.0;

    double k = 12.0;             //Strike price

    //Calculate N payoffs
    double avg = 0.0;
    for(int j=0; j < N; j++)
    {
        double temp = stoc_walk(s0, drift, volatility, days);
        if(temp > k)
        {            
            double payoff  = temp - k;
            payoff *= exp(-r);
            avg += payoff;
        }
        else
        {
            zero_trials += 1;
        }
    }

    //Average the results
    avg /= (double)N;

    //Print results
    cout << "MONTE CARLO PLAIN VANILLA CALL OPTION PRICING" << endl;
    cout << "Option price: " << avg << endl;
    cout << "Initial price: " << s0 << endl;
    cout << "Strike price: " << k << endl;
    cout << "Daily expected drift: " << drift*100 << "%" << endl;
    cout << "Daily volatility: " << volatility*100 << "%" << endl;
    cout << "Total trials: " << N << endl;
    cout << "Zero trials: " << zero_trials << endl;
    cout << "Percentage of total trials: " << zero_trials/N*100 << "%" << std::endl;

    return 0;
}

File: FinMonteCarlo.py
#!/usr/bin/python

import random
import math

def stoc_walk(p, dr, vol, periods):
    for i in range(periods):
        w = random.gauss(0.0, 1.0)
        p += dr*p + w*vol*p
    return p

s0 = 10.0
drift = 0.001502
volatility = 0.026
r = 0.02
days = 255
N = 100000
zero_trials = 0

k = 12.0

avg = 0.0

for i in range(N):
    temp = stoc_walk(s0, drift, volatility, days)
    if temp > k:
        payoff  = temp - k
        payoff *= math.exp(-r)
        avg    += payoff
    else:
        zero_trials += 1

avg = avg / float(N)

print("MONTE CARLO PLAIN VANILLA CALL OPTION PRICING")
print("Option price: ",avg)
print("Initial price: ",s0)
print("Strike price: ",k)
print("Daily expected drift: ",drift)
print("Daily expected volatility: ",volatility)
print("Total trials: ",N)
print("Zero trials: ",zero_trials)
print("Percentage of total trials: ", 100.0*float(zero_trials)/float(N))

